### ISSUE TITLE
Migrate `flutter/engine` to a Dart workspace for Dart/pub packages

### ISSUE THREAD

**Issue by matanlurey** [flutter/flutter#147883]
**tl;dr**: Dart tooling is landing support for so-called [Pub workspaces](http://flutter.dev/go/pub-workspace), the engine should use them when ready.

---

Right now, `flutter/engine` has many, many `pubspec.yaml` file each of which needs to be independently maintained, and comply with our policy of not resolving any packages to external data sources (i.e., all packages have to either be vended by the Dart SDK, or via our own `DEPS` file).

The impact of this is perhaps most profound when you look at the [necessary dependencies needed to use `package:test` as a dependency](https://github.com/flutter/engine/blob/main/testing/pkg_test_demo/pubspec.yaml), which at the time of this writing, was nearly ~50 packages (and hand-written paths that resolve to different directories depending on the package).

With upcoming [pub workspaces](http://flutter.dev/go/pub-workspace), we could instead:

- Have a single root `flutter/engine/pubspec.yaml` listing all pub dependencies we use.
- Have nested packages use `resolution: workspace`

Unblocks https://github.com/flutter/flutter/issues/133569.

See also:

- https://github.com/dart-lang/pub/issues/4127
- https://github.com/dart-lang/sdk/issues/53874
- https://github.com/dart-lang/sdk/issues/53875
